src/EnergyPlus/SwimmingPool.cc

```diff
@@ -1181,4 +1181,22 @@ Real64 MakeUpWaterVolFunct(Real64 MakeUpWaterMass, Real64 Density)
     return MakeUpWaterMass / Density;
 }
 
+void updateSwimmingPoolConditions(EnergyPlusData &state)
+{
+    // Update the temperature and flow conditions as needed to make sure that the plant gets the right temperature
+    // and so that the flow rate through the swimming pool (with an implied bypass potetially) is also correct.
+    for (auto &thisPool : state.dataSwimmingPools->Pool) {
+        auto &inletNode = state.dataLoopNodes->Node(thisPool.WaterInletNode);
+        auto &outletNode = state.dataLoopNodes->Node(thisPool.WaterOutletNode);
+        Real64 maxMassFlowRate = std::max(inletNode.MassFlowRate, outletNode.MassFlowRate);
+        inletNode.MassFlowRate = maxMassFlowRate;
+        outletNode.MassFlowRate = maxMassFlowRate;
+        if (thisPool.WaterMassFlowRate < maxMassFlowRate) {
+            Real64 bypassMassFlowRate = maxMassFlowRate - thisPool.WaterMassFlowRate;
+            outletNode.Temp = (thisPool.PoolWaterTemp * thisPool.WaterMassFlowRate + inletNode.Temp * bypassMassFlowRate) / maxMassFlowRate;
+            thisPool.WaterMassFlowRate = maxMassFlowRate;
+        }
+    }
+}
```

Comment:
Fundamentally, the equipment models should be asking the plant for a node mass flow rate. This call to SetComponentFlowRate should already be returning the proper water flow rate, at both the inlet and the outlet. 

    // Now calculate the requested mass flow rate from the plant loop to achieve the proper pool temperature
    // old equation using surface heat balance form: MassFlowRate = CpDeltaTi * ( CondTerms + ConvTerm + SWtotal + LWtotal + PeopleGain +
    // PoolMassTerm + MUWTerm + EvapEnergyLossPerArea );
    Real64 MassFlowRate = (this->WaterMass / (state.dataHVACGlobal->TimeStepSys * DataGlobalConstants::SecInHour)) *
                          ((TInSurf - TH22) / (TLoopInletTemp - TInSurf)); // Target mass flow rate to achieve the proper setpoint temperature
    if (MassFlowRate > this->WaterMassFlowRateMax) {
        MassFlowRate = this->WaterMassFlowRateMax;
    } else if (MassFlowRate < 0.0) {
        MassFlowRate = 0.0;
    }
    PlantUtilities::SetComponentFlowRate(state, MassFlowRate, this->WaterInletNode, this->WaterOutletNode, this->HWplantLoc);
    this->WaterMassFlowRate = MassFlowRate;

I think what you need here is a struct varible to keep track of what the pool mass flow rate request is this timestep. So before calling SetComponentFlowRate you would save this->WaterMassFlowRateRequest = MassFlowRate. Then call SetComponentFlowRate to update MassFlowRate, and then the difference between WaterMassFlowRateRequest and the node mass flow rate would be the bypass amount. It's seems this could be done all within the pool model instead of calling this new function multiple times from SimHVAC? And of course, I could be completely wrong on this so take this only as a suggestion. I also wonder why a pool even needs a bypass pipe since the pump sets the flow. I understand the model may want a certain flow, but it can only get what it gets.

Reply:
@rraustad Yes, I agree with you--it seems like this could all be done within the pool model.  I thought that what you suggested should have worked.  I tried that first or second I think (I tried several iterations on a similar theme) and for some reason it was not setting things correctly.  The SetComponentFlowRate call sets the flow request but that gets overwritten later in the flow resolver.  What was happening was, it would exit the pool routine with an outlet flow that matched the request and an outlet temperature that provided an energy balance at that flow.  Then, the flow resolver would come through and reset the outlet flow without touching the outlet temperature.  So, there was a net energy imbalance as a result.  Plus, the pool water flow rate would still throttle, leading to confusion for the user who logged the bug, because it didn't match the node flow rate.  I watched the flow rate and it would always get overwritten after the pool simulation step.  I also tried to query the flow lock variable within the swimming pool as a way to figure out when the flow was set, but that didn't work either.  It was at that point that I decided that I needed to take action at a higher level which led me to what I pushed.  I admit it's not ideal, but it's the only thing that I have tried that actually does what it is supposed to do.

On the topic of the pool getting a constant flow, that was also a point made by the individual who logged the bug.  The problem is that the variable flow algorithm is somewhat hard-wired into the routine.  Switching to a constant flow would involve some new equations for the main heat balance of the model.  While I understand the concern, that's more on the scope of a small development project.  Probably not a difficult one and also probably worth doing, but I'd consider it beyond a bug fix.  It would need at least one new input field to preserve the existing variable flow model.

You do make valid points and I would have guessed that what you proposed (like I said, one of the things that I tried early on) would have worked.

Reply:
OK then, you tried doing things the correct way and that did not provide a solution. So be it. I do think that what you describe points to a bug in the plant flow resolver where the plant is updating nodes without components being aware of that change.

Reply:
That there could be a plant flow resolver bug is definitely a possibility.  My two thoughts were (1) wouldn't we see problems with other components too (it would seem like a swimming pool wouldn't be the only thing to see such a problem--hence my thought that it was something specific to the pool which uniquely bridges between the zone heat balances and a plant loop) and (2) the flow resolver seems pretty solid (I mean, it's been heavily vetted over many years, right?). I could be wrong, too.  I just know that I tried a variety of things that should have worked and none of those previous attempts did anything.

Reply:
@rraustad By the way, thanks for looking at this!

Reply:
The other thing to be concerned with is setting an inlet or outlet node of a plant component that is inside a plant loop. Simply changing a node condition does not necessarily balance the entire plant. It might look good for the pool nodes, but are node conditions balanced througout the rest of the plant?

Reply:
Wow--well, that's a really good call.  What I thought was working from the perspective of the pool nodes is great.  I thought I checked the other nodes in my various test cases.  Apparently not.  The flows are fine, but the temperatures didn't get updated properly.  So, this isn't working yet...